qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_single_crystal_elastic/script_generator/elastic_single_crystal_script_generator_model.py

```diff
@@ -0,0 +1,180 @@
+"""
+
+from mantidqtinterfaces.dns_powder_elastic.data_structures.dns_elastic_powder_dataset import DNSElasticDataset
+from mantidqtinterfaces.dns_powder_tof.helpers.list_range_converters import get_normalisation
+from mantidqtinterfaces.dns_powder_tof.script_generator.common_script_generator_model import DNSScriptGeneratorModel
+import numpy as np
+from mantid.simpleapi import mtd
+
+
+class DNSElasticSCScriptGeneratorModel(DNSScriptGeneratorModel):
+    def __init__(self, parent):
+        super().__init__(parent)
+        self._data_arrays = {}
+        self._script = []
+        self._plot_list = []
+        self._sample_data = None
+        self._standard_data = None
+        self._loop = None
+        self._spacing = None
+        self._export_path = None
+        self._ascii = None
+        self._nexus = None
+        self._norm = None
+
+    def script_maker(self, options, paths, file_selector=None):
+        self._script = []
+        # shortcuts for options
+        self._vana_correction = options["corrections"] and options["det_efficiency"]
+        self._nicr_correction = options["corrections"] and options["flipping_ratio"]
+        self._sample_background_correction = options["corrections"] and options["subtract_background_from_sample"]
+        self._background_factor = options["background_factor"]
+        self._ignore_vana = str(options["ignore_vana_fields"])
+        self._sum_sf_nsf = str(options["sum_vana_sf_nsf"])
+        self._corrections = self._sample_background_correction or self._vana_correction or self._nicr_correction
+        self._export_path = paths["export_dir"]
+        self._ascii = paths["ascii"] and paths["export"] and bool(self._export_path)
+        self._nexus = paths["nexus"] and paths["export"] and bool(self._export_path)
+        self._norm = get_normalisation(options)
+        self._setup_sample_data(paths, file_selector)
+        self._set_loop()
+        error_message = ""
+        # starting writing script
+        self._add_lines_to_script(self._get_header_lines())
+        self._add_lines_to_script(self._get_sample_data_lines())
+        self._add_lines_to_script(self._get_param_lines(options))
+        self._add_lines_to_script(self._get_binning_lines(options))
+        self._add_lines_to_script(self._get_load_data_lines())
+        save_string = self._get_save_string()
+        self._add_lines_to_script(self._get_convert_to_matrix_lines(save_string))
+        return self._script, error_message
+
+    def _setup_sample_data(self, paths, file_selector):
+        self._sample_data = DNSElasticDataset(data=file_selector["full_data"], path=paths["data_dir"], is_sample=True)
+        self._plot_list = self._sample_data.create_subtract()
+
+    def _set_loop(self):
+        if len(self._sample_data.keys()) == 1:
+            self._loop = f"for workspace in wss_sample['{list(self._sample_data.keys())[0]}']:"
+            self._spacing = "\n" + " " * 4
```

Comment:
Number of spaces for indentation should be stored as a (static?) variable rather than as a literal, in case somebody decides to use a different value elsewhere.